### ISSUE TITLE
Task dependency inference doesn't work with Provider.orElse

### ISSUE THREAD

**Issue by gabrielfeo** [gradle/gradle#24141]
Given a task that takes either a provider's value as input, or another task's outputs if the provider has no value:

```kotlin
val displaySpecNotOk by tasks.registering {
  val path = localSpecPath.map { File(it).absolutePath }
    .orElse(remoteSpecUrl.map { it.outputs.files.first().absolutePath })
  inputs.property("Spec path", path)
  // ...
}
```

### Expected Behavior

When `localSpecPath` has no value, then task depends on `downloadRemoteSpec`.

```
$ ./gradlew clean displaySpecNotOk --dry-run 
:clean SKIPPED
:downloadRemoteSpec SKIPPED
:displaySpecNotOk SKIPPED
```

### Current Behavior

When `localSpecPath` has no value, task is run without `downloadRemoteSpec` (task dependency isn't inferred).

```
$ ./gradlew clean displaySpecNotOk --dry-run 
:clean SKIPPED
:displaySpecNotOk SKIPPED
```

### Context

I ran into the issue implementing similar logic and noticing the dependency task was silently not running. An easy workaround was avoiding `orElse` and checking for a value manually: https://github.com/gabrielfeo/gradle-enterprise-api-kotlin/blob/6481612e986fcf5bbdf54fa9ff5433e5d1fcdf26/build.gradle.kts#L40

#15453 seems related. Looking at [`ConfigurationCacheBuildOptionsIntegrationTest`](https://github.com/gradle/gradle/blob/44bfa93338c347a337939330bd2217185c039922/subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheBuildOptionsIntegrationTest.groovy#L36), it seems this might be a known limitation, but there's no mention in the [Incremental Build docs](https://docs.gradle.org/current/userguide/incremental_build.html#sec:inferred_task_dependencies) and the issue occurs even without configuration caching enabled.

### Steps to Reproduce 

https://github.com/gabrielfeo/gradle-task-dependencies-not-inferred-sample

### Your Environment

- Build scan URL: https://scans.gradle.com/s/ybf27zcgjkyj2 (`--dry-run`)
- Build scan URL: https://scans.gradle.com/s/kjsxcy5acns3g

**Comment by ov7a** [comment]
Thank you for your interest in Gradle!

This issue needs a decision from the team responsible for that area. They have been informed, response time may vary.

***
This is either a documentation issue (`zip` should be used instead of `orElse` here) or a valid feature request to support that. It's up to execution team to decide.

**Comment by ljacomet** [comment]
To  clarify:
* Documentation issue if we consider that having conditional task dependencies is not something we want in Gradle. Then the user leverages `zip` but both task dependencies will be executed.
* We intend to add conditional task dependency support, then this is a feature request indeed.

**Comment by kelemen** [comment]
Just a couple of notes from my report (which was closed as a duplicate of this).

Though the reporter of this issue is using Gradle properties to replicate the issue, I want to make it explicit that Gradle properties are a very important use-case in specific, because often times you want to be able to override certain parameters from the command line (via `-P`) or just plain `gradle.properties` entry. In my very specific case, I wanted to set the main class of a `JavaExec` task (which otherwise would have been auto detected from the output class files).

Also, see my simplified [build.gradle.kts](https://gist.github.com/kelemen/415443a075dbd728b3dc3dfc6237e06c) example (which works with an arbitrary `settings.gradle.kts`).

Also note, that the best workaround I have found is to check for `Provider.isPresent` (since it is not a big deal for a Gradle property). See the comment in my example linked above.

**Comment by kelemen** [comment]
Also, just a note: You have mentioned that it might be potentially classified as documentation issue where you want to ask people to move to `zip`. However, if that is what you believe should happen, then `orElse` should behave like `zip` in my opinion (as far as task dependencies go), otherwise the current behavior is very surprising, and not trivial to track down, if you don't already know what to look for (in which case the documentation is whatever).

**Comment by abstratt** [comment]
This is currently blocked by #33045
